systemtest/infra_kafka.go

```diff
@@ -21,233 +21,203 @@ import (
-	// If PortForward is set to true, and the deployment_type is k8s, start
-	// forwarding traffic to/from kafka to localhost.
-	if cfg.PortForward && dt == `"k8s"` {
-		var err error
-		runOnce.Do(func() {
-			// NOTE(marclop) These assume terraform uses the strimzi operator.
-			if cfg.PortForwardResource == "" {
-				cfg.PortForwardResource = fmt.Sprintf("%s-kafka-bootstrap", cfg.Name)
-			}
-			if cfg.PortForwardMapping == "" {
-				cfg.PortForwardMapping = "9093:9093"
-			}
-			err = portforwardKafka(
-				ctx,
-				cfg.Namespace,
-				cfg.Name,
-				cfg.PortForwardResource,
-				cfg.PortForwardMapping,
-			)
-		})
-		if err != nil {
-			return fmt.Errorf("failed to run port forward: %w", err)
-		}
+
+	// Install Strimzi using Helm.
+	if err := execCommand(ctx,
+		"helm", "repo", "add", "--force-update", "strimzi", "https://strimzi.io/charts/",
+	); err != nil {
+		return fmt.Errorf("failed to add Strimzi chart repo: %w", err)
+	}
+	if err := execCommand(ctx,
+		"helm", "upgrade", "--install", "--wait",
+		"--create-namespace", "--namespace", kafkaNamespace,
+		"strimzi", "strimzi/strimzi-kafka-operator",
+	); err != nil {
+		return fmt.Errorf("failed to install Strimzi chart: %w", err)
+	}
+
+	// Create Kafka cluster.
+	if err := execCommand(ctx,
+		"helm", "upgrade", "--install", "--wait",
+		"--namespace", kafkaNamespace,
+		"--set", "name="+kafkaClusterName,
+		"--set", "namespace="+kafkaNamespace,
+		kafkaClusterName, "../infra/k8s/kafka",
+	); err != nil {
+		return fmt.Errorf("failed to create Kafka cluster: %w", err)
+	}
+	logger().Info("waiting for Kafka cluster to be ready...")
+	if err := execCommand(ctx,
+		"kubectl", "--namespace", kafkaNamespace,
+		"wait", "--timeout=240s",
+		"--for=condition=Ready=True",
+		"kafka", kafkaClusterName,
+	); err != nil {
+		return fmt.Errorf("error waiting for Kafka broker to be provisioned by Strimzi: %w", err)
 	}
+
 	logger().Info("Kafka infastructure fully provisioned!")
 	return nil
 }
 
-func portforwardKafka(ctx context.Context, ns, name, service, mapping string) error {
-	// Patch zone for Kafka client calls to be correctly resolved.
-	srv, err := mockdns.NewServerWithLogger(map[string]mockdns.Zone{
-		// NOTE(marclop) Assumes terraform uses the strimzi operator.
-		fmt.Sprintf("%s-kafka-0.%s-kafka-brokers.%s.svc.", name, name, ns): {
-			A: []string{"127.0.0.1"},
-		},
-	}, log.New(io.Discard, "", 0), false)
-	if err != nil {
-		return err
-	}
-	if mockResolver == nil {
-		resolver := new(net.Resolver)
-		srv.PatchNet(resolver)
-		mockResolver = resolver
+// CreateKafkaTopics interacts with the Kafka broker to create topics,
+// deleting them when the test completes.
+func CreateKafkaTopics(ctx context.Context, t testing.TB, topics ...apmqueue.Topic) {
+	manager, err := NewKafkaManager(t)
+	require.NoError(t, err)
+	defer manager.Close()
+	err = manager.CreateTopics(ctx, topics...)
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		manager, err := NewKafkaManager(t)
```

Comment:
Would it make sense to just create a single kafka manage3r and only close it after the topics are deleted in the `t.Cleanup` function?

Reply:
Yes, that would be better. I'll create two cleanups: one to close, one to delete topics.